qiskit/primitives/backend_estimator_dev.py

```diff
@@ -0,0 +1,612 @@
+
+    def _postprocess_single(
+        self,
+        counts_bundle: Sequence[Counts],
+        metadata_bundle: Sequence[dict[str, Any]],
+    ) -> tuple[float, float, int]:
+        """Postprocess single counts and metadata bundles to expval, variance, and shots."""
+        expval = 0.0
+        var = 0.0
+        for counts, metadata in zip(counts_bundle, metadata_bundle):
+            paulis = metadata["paulis"]
+            coeffs = metadata["coeffs"]
+            expvals, variances = self._compute_expvals_and_variances(counts, paulis)
+            expval += np.dot(expvals, coeffs)
+            var += np.dot(variances, np.array(coeffs) ** 2)
+        shots = sum(counts_bundle[0].values())  # TODO: not correct -> counts.shots (?)
+        return expval, var, shots
+
+    ################################################################################
+    ## COMPUTATION
+    ################################################################################
+    @classmethod
+    def _compute_expvals_and_variances(
+        cls, counts: Counts, paulis: PauliList
+    ) -> tuple[tuple[float, ...], tuple[float, ...]]:
+        """Return tuples of expvals and variances for input Paulis.
+
+        Note: All non-identity Pauli's are treated as Z-paulis, assuming
+        that basis rotations have been applied to convert them to the
+        diagonal basis.
+        """
+        pairs = (cls._compute_expval_variance_pair(counts, pauli) for pauli in paulis)
+        return tuple(zip(*pairs))
+
+    @classmethod
+    def _compute_expval_variance_pair(cls, counts: Counts, pauli: Pauli) -> tuple[float, float]:
+        """Return an expval-variance pair for the given counts and pauli.
+
+        Note: All non-identity Pauli's are treated as Z-paulis, assuming
+        that basis rotations have been applied to convert them to the
+        diagonal basis.
+        """
+        shots: int = 0
+        expval: float = 0.0
+        for bitstring, freq in counts.items():
+            coeff = cls._measurement_coefficient(bitstring, pauli)
+            expval += coeff * freq
+            shots += freq
+        expval /= shots or 1  # Avoid division by zero errors if no counts
+        variance = 1 - expval**2
+        return expval, variance
+
+    @classmethod
+    def _measurement_coefficient(cls, bitstring: str, pauli: Pauli) -> int:
+        """Compute measurement coefficient from measured bitstring and target Pauli."""
+        measurement = int(bitstring, 2)
+        int_mask = cls._pauli_integer_mask(pauli)
+        return (-1) ** cls._parity_bit(measurement & int_mask, even=True)
+
+    @classmethod
+    def _pauli_integer_mask(cls, pauli: Pauli) -> tuple[int]:
+        """Build integer masks for input Pauli.
+
+        This is an integer representation of the binary string with a
+        1 where there are Paulis, and 0 where there are identities.
+        """
+        pauli_mask: list[bool] = pauli.z | pauli.x
```

Comment:
`pauli_mask` is not a list but `np.ndarray`.

Reply:
Great catch! Thanks!